src/msg/async/ProtocolV2.h

```diff
@@ -98,6 +98,7 @@ class ProtocolV2 : public Protocol {
 
   boost::container::static_vector<ceph::msgr::v2::segment_t,
 				  ceph::msgr::v2::MAX_NUM_SEGMENTS> rx_segments_desc;
+  ceph::msgr::v2::FrameAssembler tx_frame_asm;
```

Comment:
I don't think it matters whether `FrameAssembler` is initialized before `session_stream_handlers` or after because the constructor just records the pointer.  `session_stream_handlers` must be initialized before `FrameAssembler` is actually used, of course.

Ideally, `session_stream_handlers` would be a private detail of `FrameAssembler` with the latter providing just `upgrade_to_secure(connection_secret)` and `reset_security()` methods.  But `ProtocolV2` needs considerable massaging for that to happen, particularly around `reuse_connection()`...

Reply:
I think the difference is just for a human reader. If it's initialized after, there is one less potential issue to worry about/verify.

> Ideally, `session_stream_handlers` would be a private detail of `FrameAssembler` with the latter providing just `upgrade_to_secure(connection_secret)` and `reset_security()` methods.  But `ProtocolV2` needs considerable massaging for that to happen, particularly around `reuse_connection()`...

Yup, the `reuse_connection()` is pretty complicated and error-prone. With the extra memory indirection (pointer to the pair of `std::unique_ptr` instances) `FrameAssembler` avoids messing with this area. This is OK. After getting this merged, we might want to further clean the code but let's keep that separated.